python/dgl/graph.py

```diff
@@ -1101,6 +1101,64 @@ def add_edges(self, u, v, data=None):
             self._msg_index = self._msg_index.append_zeros(num)
         self._msg_frame.add_rows(num)
 
+    def del_nodes(self, vids):
+        """Remove multiple nodes.
+
+        Parameters
+        ----------
+        vids: list, tensor
+            The id of nodes to remove.
+        """
+        if self.is_readonly:
+            raise DGLError("del_nodes is not supported by read-only graph.")
+        induced_nodes = utils.set_diff(utils.toindex(self.nodes()), utils.toindex(vids))
+        sgi = self._graph.node_subgraph(induced_nodes)
+
+        if isinstance(self._node_frame, FrameRef):
+            self._node_frame = FrameRef(Frame(self._node_frame[sgi.induced_nodes]))
+        else:
+            self._node_frame = FrameRef(self._node_frame, sgi.induced_nodes)
+
+        if isinstance(self._edge_frame, FrameRef):
+            self._edge_frame = FrameRef(Frame(self._edge_frame[sgi.induced_edges]))
+        else:
+            self._edge_frame = FrameRef(self._edge_frame, sgi.induced_edges)
+
+        # FIXME(zihao): the following impl is not efficient, but directly reusing the handle of subgraph index would
```

Comment:
@jermainewang and @zheng-da could you please have a look at this, after discussion with @BarclayII we think the problem resides in "GraphInterface.Reset".

Reply:
You cannot let two GraphIndex object share the same internal handle. You will get double free problem. The reason is that there is no refcount for the handle (which is essentially a pointer to a C object). We rely on python's native object refcount to delete the handle properly. Checkout L37-40 in `graph_index.py`. The handle is deleted in the `__del__` function, which will be called **only** when the GraphIndex object is destroyed. Python makes sure that if a GraphIndex object is shallow-copied to another variable, only one `__del__` function call will be made.

If you create a new GraphIndex and reuse the internal handle of an existing GraphIndex object, two `__del__` function will delete the same handle.

Reply:
My solution:
* Change `graph_index.SubgraphIndex` to **not** inherit `GraphIndex`. Actually, our C++ data structure is like this. The subgraph index only contains the graph structure, the parent graph, induced nodes and edges. As a result, the object returned by `node_subgraph` is not necessarily readonly graph or not.
* In `del_nodes/edges` functions, directly set the returned graph structure as the `_graph` field in the `DGLGraph` object.